python/tk_multi_publish2/dialog.py

```diff
@@ -230,23 +323,160 @@ def _update_details_from_selection(self):
+            self._clear_custom_settings_page()
+            self._current_tasks = new_task_selection
+            return
+
+        # A task was picked, so make sure our page is in foreground.
+        self.ui.details_stack.setCurrentIndex(self.TASK_DETAILS)
+
+        # If the new task selection does not have a custom UI, a simple tear down of the custom UI
+        # and setting the built-in fields will suffice.
+        if not new_task_selection.has_custom_ui:
+            logger.debug("Clearing custom UI and using default task details...")
+            self._clear_custom_settings_page()
+
+            # All the items are of the same type,
+            self.ui.settings_stack.setCurrentIndex(self.BUILTIN_TASK_DETAILS)
+            self.ui.task_icon.setPixmap(new_task_selection.plugin.icon)
+            self.ui.task_name.setText(new_task_selection.plugin.name)
+            self.ui.task_description.setText(new_task_selection.plugin.description)
+
+            # skip settings for now
+            # self.ui.task_settings.set_data(task.settings.values())
+
+            self._current_tasks = new_task_selection
+            return
+
+        # At this point we can assume we're going to have to show a UI, because new task exists
+        # and it has a custom UI.
+        self.ui.settings_stack.setCurrentIndex(self.CUSTOM_TASK_DETAILS)
+
+        # Now figure out if we need to create/replace the widgets.
+        if (
+            # If we didn't have a current task before or...
+            not self._current_tasks or
+            # We did have one, but it wasn't the same task type as the current one
+            not self._current_tasks.is_same_task_type(new_task_selection)
+        ):
+            logger.debug("Building a custom ui for %s.", new_task_selection.plugin)
+            self._clear_custom_settings_page()
+            widget = new_task_selection.plugin.run_create_settings_widget(self.ui.custom_settings_page)
+            self.ui.custom_settings_page.layout().addWidget(widget)
+        else:
+            logger.debug("Reusing custom ui from %s.", new_task_selection.plugin)
+            # Same plugin type, we can simply fetch back the widget
+            widget = self.ui.custom_settings_page.layout().itemAt(0).widget()
+
+        # Update the UI with the settings from the current plugin.
+        if self._push_settings_into_ui(new_task_selection, widget):
+            # Alright, we're ready to deal with the new task.
+            self._current_tasks = new_task_selection
+        else:
+            self._current_tasks = self.TaskSelection()
+
+    def _pull_settings_from_ui(self, selected_tasks):
+        """
+        Retrieves settings from the UI and updates the task's settings.
+
+        :param selected_tasks: A :class:`TaskSelection` of tasks to update based
+            on the values edited in the UI.
+        """
+        if selected_tasks.has_custom_ui:
+            widget = self.ui.custom_settings_page.layout().itemAt(0).widget()
```

Comment:
Feels like we should have a widget that exposes an API for retrieving the contained, custom widget. Something like: 

```python
widget = self.ui.custom_settings_page.get_custom_settings_widget()
```